lua/nvim-treesitter/textobjects/select.lua

```diff
@@ -11,17 +11,22 @@ local M = {}
 function M.select_textobject(query_string, keymap_mode)
   local bufnr, textobject = shared.textobject_at_point(query_string)
   if textobject then
-    -- Detect if selection should be made linewise depending keymaping mode.
-    -- This should be done before calling `ts_utils.update_selection()` because
-    -- it forces 'v' mode.
-    local force_linewise = (keymap_mode == "x" and vim.fn.visualmode() == "V") or
-      (keymap_mode == "o" and vim.fn.mode(1) == "noV")
-    -- Make regular selection
-    ts_utils.update_selection(bufnr, textobject)
-    -- Possibly make selection linewise
-    if force_linewise then
-      vim.fn.nvim_exec("normal! V", false)
+    local selection_mode = "charwise"
+
+    -- Update selection mode based on keymapping mode
+    local ctrl_v = vim.api.nvim_replace_termcodes("<C-v>", true, true, true)
+    if keymap_mode == "x" then
+      local t = { V = "linewise"}
+      t[ctrl_v] = "blockwise"
+      selection_mode = t[vim.fn.visualmode()]
+    elseif keymap_mode == "o" then
+      local t = { noV = "linewise" }
+      t["no" .. ctrl_v] = "blockwise"
+      selection_mode = t[vim.fn.mode(1)]
```

Comment:
Couldn't the util function do all this complicated mapping? The util function could also accept "<c-v>" and all these cryptic abbreviations for all the modes.

Reply:
I can wrap these lines in separate function, like `M.detect_selection_mode(keymap_mode)`.

But I don't really understand the second part. Those abbreviations are coming from Vim's (well Neovim's) functionality. There are only three types of selections (charwise, linewise, blockwise) which can be detected in certain ways depending on keymapping mode. I can extend that utility function to also acknowledge "v" (Visual + Selection) and "s" (Selection) modes (with the same treatment as "x"), but this seems to be it. General mapping with "" will produce charwise selection.

Reply:
Just pushed what I see the reasonable utility function. It defaults to "charwise" selection mode which might be updated on certain supported keymapping modes.

Reply:
This can now be simplified with the updated upstream function, right?

Reply:
Sure. I like this "symmetric" treatment of visual and operator-pending mode more, but I guess it has minor execution speed issues.

Reply:
ok, so this is ready to be merged?

Reply:
Are you familiar with this issue? https://github.com/nvim-treesitter/nvim-treesitter-textobjects/issues/39

Reply:
> ok, so this is ready to be merged?

Just now spotted the 4-space indent instead of 2-space and pushed fix. Otherwise, this does what is discussed: adds possibility to make linewise and blockwise selection (during visual selection and in operator-pending mode) in the best way I can currently suggest.

> Are you familiar with this issue? #39

I've looked at it. Not an expert in intricacies of Vim's binding mechanisms, but the core of a problem seems to lie in logic of both 'ts_utils.update_selection()` and `TSTextobjectSelect`. `ts_utils.update_selection()` currently makes selection by sequentially moving to start of textobject, executing `v`, `V`, or `C-v`, and moving to end of textobject. At the same time `TSTextobjectSelect` accepts only one argument (which is query), so selection will always be done in with `v` command (default). So with enabled `<cmd>` mapping, typing `vaf` executes: `v`, go to start, `v` (which cancels first v), to the end. This looks like cursor just moves to the end of textobject.

However, I don't think this has something to do with this PR (which works with suggested method of mapping through `keymaps` option of `'nvim-treesitter.configs'.setup`).